### ISSUE TITLE
Wazuh packages redesign tier 2

### ISSUE THREAD

**Issue by havidarou** [wazuh/wazuh-packages#2904]
# Description

**Wazuh-packages (QA, Agent, CppServer)**
- Clean up everything
- Add support to PPC packages (Agent)

## Requirements and Restrictions

- [x] Wazuh packages redesign tier 1 requirements should be taken into account
- [ ] The new repository must include QA related code and tools
  - [x] New Jenkins should be used for everything QA related after 5.0
  - [x] Current wazuh-packages repository is deprecated after 4.10
  - [ ] Current wazuh-jenkins repository is deprecated after 4.10
  - [ ] Current wazuh-tools repository is deprecated after 4.10: https://github.com/wazuh/wazuh-qa-automation/issues/417
  - [x] Migrate scripts and pipelines that sync GitHub projects with GitHub.(QA)
  - [x] Identify existing [pipelines](https://docs.google.com/spreadsheets/d/1gGbqjJwcevlq6ICdYN-F0tP45ooeOZyZ/edit?usp=sharing&ouid=110350653643919800893&rtpof=true&sd=true) and migrate those that continue to provide coverage to Wazuh 4.x.  (QA)
- [x] The release procedure should be improved by:
       - Improving repository metadata generation (avoid full metadata sync)
       - Reducing execution time
       - Incorporating backups and storage
       - Mirroring production bucket (release to the mirror, sync main with the mirror, restore main with the mirror)
- [x] The following will be migrated to the new DevOps repositories (both packages and automation):
       - OVA
       - AMI
       - Wazuh unattended scripts
       - Wazuh Puppet
       - Wazuh Docker
       - Wazuh Kubernetes
       - Wazuh Ansible
       - Training environment
       - Demo environment
- [x] Add support to PPC packages (Agent)
- [x] The following installers will be migrated from wazuh-packages repository: 
       Agent installers:
        - HP-UX
        - Solaris Intel/SPARC
        - AIX

## Auditor

- Involved: DevOps, QA, Agent, CppServer
- esponsible: @wazuh/devel-qa-div3
- DRI name: @davidjiglesias
- Objective: Wazuh packages redesign tier 2

----

# Planning

<details><Summary>Precondition</Summary>

### Previous tier

- https://github.com/wazuh/internal-devel-requests/issues/187

### Dependencies

- https://github.com/wazuh/wazuh-jenkins/issues/5927
- https://github.com/wazuh/wazuh-jenkins/issues/6357
- https://github.com/wazuh/wazuh-jenkins/issues/6471

</details>

<details><Summary>Spike</Summary>

- https://github.com/wazuh/wazuh-qa/issues/5527 (QA)
- https://github.com/wazuh/wazuh-qa/issues/5574 (QA)
- https://github.com/wazuh/wazuh-qa/issues/5599
</details>

<details><Summary>Tasks</Summary>

### QA

- https://github.com/wazuh/wazuh-jenkins/issues/6811
- https://github.com/wazuh/wazuh-jenkins/issues/6770
- https://github.com/wazuh/wazuh-qa-automation/issues/56
- https://github.com/wazuh/wazuh-qa-automation/issues/26 
- https://github.com/wazuh/internal-devel-requests/issues/1558
- https://github.com/wazuh/internal-devel-requests/issues/1555
- https://github.com/wazuh/wazuh-qa-automation/issues/113
- https://github.com/wazuh/wazuh-qa-automation/issues/111
- https://github.com/wazuh/wazuh-qa-automation/issues/160
- https://github.com/wazuh/wazuh-qa/issues/5556 
- https://github.com/wazuh/wazuh-qa-automation/issues/130
- https://github.com/wazuh/wazuh-qa-automation/issues/84

**Further improvements are required**
- https://github.com/wazuh/wazuh-qa-automation/issues/221
- https://github.com/wazuh/wazuh-qa-automation/issues/226
- https://github.com/wazuh/wazuh-qa-automation/issues/295

**Move to 4.10.1**
- https://github.com/wazuh/wazuh-qa-automation/issues/308
- https://github.com/wazuh/wazuh-qa-automation/issues/309

### DevOps

- https://github.com/wazuh/wazuh-automation/issues/1658 
- https://github.com/wazuh/wazuh-virtual-machines/issues/22
- https://github.com/wazuh/wazuh-installation-assistant/issues/21 
- https://github.com/wazuh/wazuh-automation/issues/1766
- https://github.com/wazuh/wazuh-automation/issues/1767 
- https://github.com/wazuh/wazuh-qa/issues/5654
- https://github.com/wazuh/wazuh-automation/issues/1873
- https://github.com/wazuh/internal-devel-requests/issues/1715

### Agent

- https://github.com/wazuh/wazuh-agent-packages/issues/55
- https://github.com/wazuh/wazuh/issues/26490
- https://github.com/wazuh/wazuh/issues/25683 

</details>

---

# Conclusion (WIP) 

## QA

- Coverage: [here](https://docs.google.com/spreadsheets/d/1ic9Sk3qehfqRZ7-_hqdkPfjHcT57VP1VXRc6RBB0at8/edit?usp=sharing)

<details><Summary> Release procedure </Summary>

 - [S3 sign check](https://github.com/wazuh/wazuh-jenkins/issues/6690): S3 sign checking now works in multi-threaded mode, where each thread runs in a separate VM to handle its errors (destroying the VM and leaving a log of the error). It helps achieve asynchronous independence from the executions in S3_sign_check. 
 - [Reducing execution time](https://github.com/wazuh/wazuh-jenkins/issues/6813): There's no way to speed up DEB metadata generation without creating bigger problems that could compromise the process.
- [Allow release-tool.py execution on AWS](https://github.com/wazuh/wazuh-jenkins/issues/6810): It allows the launch tool to run on AWS infrastructure (4vcpu and 8GB). 
In addition, it was adapted to local execution and in local containers, providing a single entry point for the tool. It balanced the increase in time in metadata generation. 
  | Stage |  v4.9.0-beta2 | Current implementation on AWS|
  |---|---|---|
  | RPM - Download| 52m | 7m|
  | RPM - Metadata and upload|-|4m20s|
  | RPM - Metadata|20s|-|
  | RPM - Upload|40s|-|
  | DEB - Download| 49m | 5m|
  | DEB - Metadata and upload|-|46m|
  | DEB - Metadata|22m|-|
  | DEB - Upload|40s|-|
- [Improving repository metadata generation](https://github.com/wazuh/wazuh-jenkins/issues/6812): Added new documentation on how to modify recipe files and regenerate metadata without adding files to the repository, added new use cases such as Stage support, release protocol, and metadata regeneration, improved requirements, and removed redundant information. 
- [Incorporating backups and storage](https://github.com/wazuh/wazuh-jenkins/issues/6814): It provides an optional backup mechanism and a precondition feature that allows the execution of ordered tasks and allows some operations before starting with the core release procedure.
- [Mirroring production bucket (release to the mirror, sync main with the mirror, restore main with the mirror)](https://github.com/wazuh/wazuh-jenkins/issues/6815): Currently, we sync prod into prod-mirror to guarantee that prod-mirror is a mirror and deploy packages into prod-mirror. After that, the backup prod by into the directory (s3://packages.wazuh.com/Backups/Backup_{timestamp}/{major}.x) and we sync prod-mirror into prod. Finally, we AWS S3-to-S3 copy, without localhost intervention, only deltas, and then we invalidate the cache of the prod. 
</details>

<details><Summary> Restore support to PPC packages (Agent - deb ppc64el, rpm ppc64le) </Summary>

The changes consisted of:
- Update Wazuh documentation.
- Adapt release procedures to support PPC packages (package generation, package signing and verification, and release tool deployment).
- Update release issues (tests).
</details>

<details><Summary> The new repository must include QA related code and tools </Summary> 

- Defines a private repository: “wazuh-qa-automation”. 
- Currently, we have two branches: 4.10.0 and the main.
- The new repository includes:
  - New templates for reporting bugs, requesting new support, and creating PRs.
  - New README with a use guide for the tools migrated.
  - Apply new structure and good practices. 
  - The 4.10.0 and 5.0.0 branches have the code migrated. 
  - For 4.10.0
    - Workflow 
       - s3_sign_check: Adds the requirement for a GitHub PAT that must be set using the GITHUB_TOKEN environment variable.
       - dtt1: Adds two steps to workflow files. Now, the token must be manually filled in before execution.
    - Adapted workflows that use the Allocator module that are on wazuh-automation.
    - Identified existing pipelines to migrate.
    - Migrate JobFlow tools
    - Migrated Provision module
    - Migrated release procedures
    - Migrated new Jenkins's code
    - Migrated Tests module and DTT workflows
    - Migrated the environments provisioning module
    - Migrated the System tests - All
    - Migrated the End to End tests - Only VD
    - Migrated scripts and pipelines that sync GitHub projects with GitHub.    
    - Migrated pipelines of: 
      - Test_registration
      - Test Register Tier
      - Test Service
      - Test_upgrade pipeline to the new Jenkins
      - Test_upgrade_tier pipeline to the new Jenkins
      - Test_stress pipeline to the new Jenkins.
      - Test_stress_gh pipeline to the new Jenkins.
      - Test_stress_Vagrant pipeline to the new Jenkins.
      - Test_stress_tier pipeline to the new Jenkins.
      - Procedure_ecs_task pipeline to the new Jenkins.
      - Procedure_ecs_task_tier pipeline to the new Jenkins.
      - procedure_cicd_release_issues pipeline to the new Jenkins.
      - Procedure_gh_project_syncup pipeline to the new Jenkins.
    - Delete any dependency of wazuh-packages repository .
    - Delete any dependency of  wazuh-tools repository .
    - Delete any dependency of wazuh-packages repository .
    - Delete any dependency of qa-integration-framework repository .
    - Delete any dependency of qa-system-framework repository .
    - Delete any dependency of wazuh-packages repository . 
    
</details>

<details><Summary> The installers migrated from wazuh-packages repository</Summary>

- DevOps GHA was adapted with a best-effort approach but a unique interface should be defined and populated among all subsystems to be releasable QA release procedures were too related to Wazuh Packages Redesign Tier 1, difficult to modularize and extend. Following aspects are some of them: 
  - Generate packages only consider deliverable as files. This is not true for i.e AMI and Docker
  - Generate packages only consider one file per target. This is not true for installation-assistant
  - Generate packages set several constraints on GHA that are not documented: id, id as part of the name, checksum
  - Sign check only allows tagging files that imply an upload procedure. This is not necessary when the file does not need a signature

</details>

---

<details><Summary>OUT OF SCOPE: "Packages Redesign Tier 2"</Summary>

- We are working on https://github.com/wazuh/wazuh-qa-automation/issues/42 which proposes new processes and practices to work.
- The 5.0.0 branch should remove everything that Migrate has deprecated.
- Once the target tasks are complete, we will work on implementing some improvements, which are:
  <details><Summary>Release Procedures</Summary>
   
   Medium priority improvements:
   - **tools/release/src/\*\*/docker/Dockerfile_tests:** Each module's docker test image.
       We should re-evaluate if it's necessary to run each module's unit tests on a docker image, noting that it consumes time and may not add much value to the development itself.
   - **.github/workflows:** Main GHA workflows directory
       - Currently, all the Release Procedures modules' Unit Tests are being executed on GHA and running on Docker, which means that we build a Docker image and run it on each GHA, and it is resources-consuming.
       - All the GHA Workflows can be simplified in just one workflows that runs all the UTs (but first, it would be great if we get rid of the Docker build)
   
   </details>
   
   
   <details><Summary>Testing</Summary>
   
   High-priority improvements:
   - **tools/testing/src/testing/testing.py:** Main testing module file
       - The module's parameters are strongly attached to DTT, it must be more generic to re-use it on different suites.
       - It expects the target and dependencies inventories to be on a specific path with a desired format, so it won't detect the target system correctly if the inventory is in a custom directory.
   - **tools/testing/src/testing/playbooks/test.yml** Test execution playbook
       The test path is generated "dynamically" but it depends completely on the DTT format
   - **tests/test_functional/test_system/test_deployability/workflows:** DTT workflows directory
       - The workflows are outdated, and its using the legacy allocator path, it must be updated to use them.
       - Wazuh version on the Workflows is hardcoded using an outdated value.
       - Add parametrization on the Workflows where possible.
   - **tools/testing/tests:** Testing tool Unit Tests directory
       There are no Unit Tests developed, we must create a set of tests to ensure its correct functionality 
       Related issues:
       - https://github.com/wazuh/wazuh-qa/issues/4993
   
   Medium priority improvements:
   - **tests/test_functional/test_system/test_deployability/tests:** Main DTT tests directory
       The test helpers have a lot of duplicated code, and there are several wrongly handled conditional flows. We could include unit tests for the helpers to validate its correct behavior
   
   </details>
   
   <details><Summary>JobFlow</Summary>
   
   Medium priority improvements:
   - **tools/jobflow/README.md:** Main documentation file of the JobFlow tool
       Currently, the documentation is DTT1-specific, it does not help the user to understand what the tool does, and how it can be used for different cases.
   
   ### Low priority improvements:
   - **tools/jobflow/tests:** Unit Tests suite
       This suite uses Python's default test runner UnitTests utilities and pytest as the runner, it could be improved by only using pytest and its utilities, which is recommended.
       Related issues:
       - https://github.com/wazuh/wazuh-qa/issues/4993
   - **tools/jobflow/examples:** Workflows examples directory
       We must add more Workflows as examples of different use cases.
   
   > Other improvements could be found here: https://github.com/wazuh/wazuh-qa/issues/5044
   
   </details>
   
   <details><Summary>Provision</Summary>
    
    Medium priority improvements:
   - **tools/provision/README.md:** Main documentation file of the Provision tool
       This README is strongly related to the JobFlow, it could be improved detailing mainly the standalone execution and having the JobFlow workflow as a secondary use.
   
   ### Low-priority improvements:
   - **tools/provision/src/provision/playbooks/wazuh:** Wazuh-related playbooks
       This directory could be re-structured, probably the non install/uninstall actions (register, services, set repo) could be separated on a different folder as these actions are not specifically related to an installation type.
   - **tools/provision/tests:** Unit Tests suite
       This suite uses Python's default test runner UnitTests utilities and pytest as the runner, it could be improved by only using pytest and its utilities, which is recommended.
       Related issues:
       - https://github.com/wazuh/wazuh-qa/issues/4993
    
   </details>
   <details><Summary>Jenkins</Summary>
   
   Low-priority improvements:
   - **tools/jenkins/tests:** tool's Unit Tests directory
       Must implement more Unit Tests for this tool
   - **tools/jenkins/src/job-builder:** Jenkins pipelines as a code
       Add a README.md detailing this folder, it could be implemented here or in the main Jenkins src directory (tools/jenkins/src), in the second case, it should also includes a summary about bootstrap.
       
   </details>
</details>

---

### Deliverables

<details><Summary> 4.10.0 Alpha 2 </Summary>

### QA

- https://github.com/wazuh/wazuh-qa/issues/5556 
   - Create repository
   - Create branches: 4.10.0, 4.10.1, 5.0.0
   - Migrate JobFlow (ex Workflow engine) tool 
   - Migrate Provision module 
   - Migrate release procedures 
   - Migrate new Jenkins's code 
   - Migrate Tests module and DTT workflows
   - Migrate code from qa-integration-framework
   - Migrate code from qa-system-framework
   - Migrate code from wazuh-tools 
   - Migrate code from wazuh-qa: 
      - Environments provisioning module
      - System tests - All
      - End to End tests - Only VD
   - Migrate code from wazuh-jenkins:
      - Test_upgrade and Test_upgrade_tier
      - Procedure_ecs_task and Procedure_ecs_task_tier
      - Procedure_gh_project_syncup
      - Test_stress
      - Test_stress_gh
      - Test_stress_tier
      - Test_stress_Vagrant
   - Migrate release issue generation template
- https://github.com/wazuh/wazuh-jenkins/issues/6811
- https://github.com/wazuh/wazuh-jenkins/issues/6770
- https://github.com/wazuh/wazuh-qa-automation/issues/56
- https://github.com/wazuh/wazuh-qa-automation/issues/26 (Maintenance with all of 4.10.0-alpha1)
- https://github.com/wazuh/internal-devel-requests/issues/1558
- https://github.com/wazuh/internal-devel-requests/issues/1555
- https://github.com/wazuh/wazuh-qa-automation/issues/113
- https://github.com/wazuh/wazuh-qa-automation/issues/111

### DevOps

- https://github.com/wazuh/wazuh-automation/issues/1658 
- https://github.com/wazuh/wazuh-virtual-machines/issues/22
- https://github.com/wazuh/wazuh-installation-assistant/issues/21 
- https://github.com/wazuh/wazuh-automation/issues/1766
- https://github.com/wazuh/wazuh-automation/issues/1767 
- https://github.com/wazuh/wazuh-qa/issues/5654

### Agent

- https://github.com/wazuh/wazuh-agent-packages/issues/55

</details>

<details><Summary> 4.10.0 - Alpha 3 </Summary>

### QA

- https://github.com/wazuh/wazuh-qa-automation/issues/84
- https://github.com/wazuh/wazuh-qa-automation/issues/221
- https://github.com/wazuh/wazuh-qa-automation/issues/226
- https://github.com/wazuh/wazuh-qa/issues/5556 
   - Migrate code from wazuh-jenkins:
      - wazuh_qa_environment
      - Procedure_docker_testing_image_generation 
      - Test_e2e_system
   - Review the migrated code
   - Update the README.md
   - Maintenance of branches.
   - Create documentation basic documentation
- https://github.com/wazuh/wazuh-qa-automation/issues/274   

### Agent

- https://github.com/wazuh/wazuh-agent-packages/issues/55

</details>

<details><Summary> 4.10.1 - Alpha 1 </Summary>

### QA

- https://github.com/wazuh/wazuh-qa/issues/5556 
   - Migrate code from wazuh-jenkins:
      - workloads
      - Test_integration_endpoints   
</details>

**Comment by rauldpm** [comment]
Issue content migrated from https://github.com/wazuh/internal-devel-requests/issues/993

**Comment by davidjiglesias** [comment]
To be evaluated as part of Wazuh packages redesign tier 2

**Comment by damarisg** [comment]
If there is time, we can introduce the following: 
- https://github.com/wazuh/wazuh-qa/issues/5681
